src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs

```diff
@@ -544,7 +557,7 @@ private async Task OnBatchScopeDisposedMaybeAsync(bool useAsync)
 
             var hasAnalyzerChanges = _analyzersAddedInBatch.Count > 0 || _analyzersRemovedInBatch.Count > 0;
 
-            await _projectSystemProjectFactory.ApplyBatchChangeToWorkspaceMaybeAsync(useAsync, solutionChanges =>
+            await _projectSystemProjectFactory.ApplyBatchChangeToWorkspaceMaybeAsync(useAsync, (solutionChanges, projectUpdateState) =>
```

Comment:
this is one of the larger transformations that changed.

The basic idea here is that there is an immutable `ProjectUpdateState` that stores changes to state that happen inside the transformation function.
This is passed to the transformation function, and updated and returned from the transformation.

Once the transformation actually applies, that state is then applied as the current state and does other state updates that should happen once.

If there are multiple attempts, the `projectUpdateState` from the failed attempt is just discarded.